nixos/modules/services/monitoring/certspotter.nix

```diff
@@ -0,0 +1,126 @@
+      name = "hooks.d";
+      path = pkgs.linkFarm "certspotter-hooks" (lib.imap1 (i: path: {
+        inherit path;
+        name = "hook${toString i}";
+      }) cfg.hooks);
+    });
+in
+{
+  options.services.certspotter = {
+    enable = lib.mkEnableOption "Cert Spotter, a Certificate Transparency log monitor";
+
+    sendmailPath = lib.mkOption {
+      type = lib.types.path;
+      description = ''
+        Path to the `sendmail` binary. By default, the local sendmail wrapper is used
+        (see `config.services.mail.sendmailSetuidWrapper`).
+      '';
+      example = lib.literalExpression ''"''${pkgs.system-sendmail}/bin/sendmail"'';
+    };
+
+    watchlist = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Domain names to watch. To monitor a domain with all subdomains, prefix its name with `.` (e.g. `.example.org`).";
+      default = [ ];
+      example = [ ".example.org" "another.example.com" ];
+    };
+
+    emailRecipients = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "A list of email addresses to send certificate updates to.";
+      default = [ ];
+    };
+
+    hooks = lib.mkOption {
+      type = with lib.types; listOf path;
+      description = ''
+        Scripts to run upon the detection of a new certificate. See `man 8 certspotter-script` or [the GitHub page](https://github.com/SSLMate/certspotter/blob/master/man/certspotter-script.md) for more info.
+      '';
+      default = [];
+      example = lib.literalExpression ''
+        [
+          (pkgs.writeShellScript "certspotter-hook" '''
+            echo "Event summary: $SUMMARY."
+          ''')
+        ]
+      '';
+    };
+
+    extraFlags = lib.mkOption {
+      type = with lib.types; listOf str;
+      description = "Extra command-line arguments to pass to Cert Spotter";
+      example = [ "-start_at_end" ];
+      default = [ ];
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.watchlist != [ ];
+        message = "You must specify at least one domain for Cert Spotter to watch";
+      }
+      {
+        assertion = cfg.hooks != [] || cfg.emailRecipients != [];
+        message = "You must specify at least one hook or email recipient for Cert Spotter";
+      }
+      {
+        assertion = (cfg.emailRecipients != []) -> (cfg.sendmailPath != "/run/current-system/sw/bin/false");
+        message = ''
+          You must configure the sendmail setuid wrapper (services.mail.sendmailSetuidWrapper)
+          or services.certspotter.sendmailPath
+        '';
+      }
+    ];
+
+    services.certspotter.sendmailPath = lib.mkMerge [
+      (lib.mkIf (config.services.mail.sendmailSetuidWrapper != null) (lib.mkOptionDefault "/run/wrappers/bin/sendmail"))
+      (lib.mkIf (config.services.mail.sendmailSetuidWrapper == null) (lib.mkOptionDefault "/run/current-system/sw/bin/false"))
+    ];
+
+    users.users.certspotter = {
```

Comment:
Do we need to specify a user, or can we use systemd's DynamicUser? If we use that, we can repoint the directory that CertSpotter uses for its state to systemd-provided $STATE_DIRECTORY, so as to make the configuration slightly more standard.

Reply:
we can't use DynamicUser as I explained above (because the hooks need flexibility)

Reply:
Thanks, I missed the previous explanation (and it obviously does make sense).

Still, could we use `StateDirectory` to handle the state?

Reply:
it's already created with `createHome`, is there a benefit to using `StateDirectory`?

Reply:
One of uniformity: it makes it easier to keep track of where state is stored for all the units one has.

It also handles changes in configuration somewhat better: TTBOMK `createHome` will create it only if the user doesn't exist and will do nothing otherwise. If e.g. someone changes their configuration to run this unit with a different group, the directory will not be chgrped. `StateDirectory` OTOH will.